### PR TITLE
Implement waitForReadinessProbe in livenessProbe

### LINKED ISSUE #27114: LivenessProbe should start after ReadinessProbe Succeeded if ReadinessProbe is specified
Our initial understanding is that liveness probe will start to check after readiness probe was succeeded but it turn out not to be like that.

We are testing with our system that has a long boot time, an approximation of boot time is between 1-3 minutes. We specify readiness probe with same url of liveness probe and specify initial delay of liveness probe to be 30 seconds, we found that the pod is killed by failing of liveness probe while readiness probe still in a failure.

So, why don't we specify initial delay of liveness probe to be more than 3 minutes? Well, it might be a case when readiness probe succeed at first minute and after that the pod will fail to do a job before liveness probe will start. So, it will affect running service.

Another point is why we don't only put readiness probe and leave out liveness probe? When readiness probe fail again it might take out of service as well so, it don't cause any affect for running service but it won't be restarted and we have to do manual restart them.

There are some concerns we thought about, here are list:
- What if readiness probe never succeed while starting a pod, can we specify a maximum number of check will be performed and if readiness probe fail more than that number, we may consider that pod should be killed.
- What if someone want to set readiness probe to put the pod in maintenance mode, the pod might be killed by above point if readiness probe fail more than that number or liveness probe will kill it. Maybe, we add one more state of readiness probe called `Maintenance` so that liveness probe should stop working when enter this state and no number of failing readiness probe should not be considered.

What do you think?

### PR DESCRIPTION
**Pull request by matthyx**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**
Adds a Boolean option `waitForReadinessProbe` to livenessProbe (default is false) that waits for the pod to be ready before checking for liveness.

**Which issue(s) this PR fixes**
Fixes #27114

**Special notes for your reviewer**
It's been a long time since I last contributed to k/k, I don't know where to document this change, please point me to the correct page.
Also I need guidance for the milestone.

**Does this PR introduce a user-facing change?**
Yes.
```release-note
Added a Boolean option to livenessProbe that waits for the pod to be ready before checking for liveness.
```

### CODE REVIEW

**Review by jimethn — APPROVED** [review]

**Review by thockin — COMMENTED** [review]
I am not sure I buy the semantic here.  I will reply on the issue.

File: pkg/apis/core/types.go [hunk 1]
```diff
@@ -1850,6 +1850,10 @@ type Probe struct {
 	// Minimum consecutive failures for the probe to be considered failed after having succeeded.
 	// +optional
 	FailureThreshold int32
+	// For livenessProbe only. Wait until readinessProbe succeeds to check livenessProbe.
+	// Default is false.
+	// +optional
+	WaitForReadinessProbe bool
```
[the remark this answers is not in the document]
Comment (matthyx) [reply pkg/apis/core/types.go]: Thanks Tim!
I have thought about composition too... let's see how it makes the handling more complex.
I will wait until you settle the conversation in the issue though.

**Comment by matthyx** [comment]
Superseded by #71449

**Pull request closed without merging**